kernel/src/history_manager/search.rs

```diff
@@ -0,0 +1,253 @@
+//! This module defines the [`binary_search_by_key_with_bounds`] utility method. This can be used to
+//! search over sorted slices of values to find the greatest lower or least upper bounds.
+use std::cmp::Ordering;
+use std::error::Error;
+use std::fmt::Debug;
+
+use delta_kernel_derive::internal_api;
+
+/// Defines the type of bound to return from a binary search operation.
+///
+/// For a search operation over `values` using key `key`:
+/// * [`Bound::LeastUpper`] - Finds the smallest index `i` such that `values[i] >= key`.
+///   This represents the first element greater than or equal to the search key.
+///
+/// * [`Bound::GreatestLower`] - Finds the largest index `i` such that `values[i] <= key`.
+///   This represents the last element less than or equal to the search key.
+#[internal_api]
+#[allow(unused)]
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum Bound {
+    LeastUpper,
+    GreatestLower,
+}
+
+#[internal_api]
+#[allow(unused)]
+#[derive(Debug)]
+pub(crate) enum SearchError<T: Error> {
+    OutOfRange,
+    KeyFunctionError(T),
+}
```

Comment:
any reason to not use `thiserror`?

Reply:
(probably want docs + impl Error regardless)

Reply:
I'd intended for the cases to be handled by the caller. Using `thiserror` lets the caller propagate it up. I can move it over to `thiserror` if you feel strongly tho.

Note: there's some precedence: Rust's binary search function returns `Result<usize, usize>`, and usize doesn't implement `Error`  The user should handle out of range.